tezos/new_context/src/timings.rs

```diff
@@ -49,10 +69,545 @@ pub fn context_action(
+
+        current.replace((id.to_string(), hash));
+
+        Ok(())
+    }
+
+    fn get_id_on_table(
+        sql: &Connection,
+        table_name: &str,
+        hash_string: &str,
+    ) -> Result<Option<i64>, SQLError> {
+        let mut stmt = sql.prepare(&format!(
+            "SELECT id FROM {table} WHERE hash = ?1;",
+            table = table_name
+        ))?;
+        let mut rows = stmt.query([hash_string])?;
+
+        if let Some(row) = rows.next()? {
+            return Ok(Some(row.get(0)?));
+        };
+
+        Ok(None)
+    }
+
+    fn insert_checkout(
+        &mut self,
+        context_hash: ContextHash,
+        irmin_time: f64,
+        tezedge_time: f64,
+    ) -> Result<(), SQLError> {
+        self.checkouts.add(irmin_time, tezedge_time);
+        self.set_current_context(context_hash)?;
+        self.insert_action(&Action {
+            name: "checkout".to_string(),
+            key: vec![],
+            irmin_time,
+            tezedge_time,
+        })
+    }
+
+    fn insert_commit(&mut self, irmin_time: f64, tezedge_time: f64) -> Result<(), SQLError> {
+        self.commits.add(irmin_time, tezedge_time);
+        self.insert_action(&Action {
+            name: "commit".to_string(),
+            key: vec![],
+            irmin_time,
+            tezedge_time,
+        })?;
+        self.update_global_stats()
+    }
+
+    fn insert_action(&mut self, action: &Action) -> Result<(), SQLError> {
+        let block_id = self.current_block.as_ref().map(|(id, _)| id.as_str());
+        let operation_id = self.current_operation.as_ref().map(|(id, _)| id.as_str());
+        let context_id = self.current_context.as_ref().map(|(id, _)| id.as_str());
+
+        let key = if action.key.is_empty() {
+            None
+        } else {
+            Some(action.key.join("/"))
+        };
+
+        self.sql.execute(
+            "
+        INSERT INTO actions
+          (name, key, irmin_time, tezedge_time, block_id, operation_id, context_id)
+        VALUES
+          (:name, :key, :irmin_time, :tezedge_time, :block_id, :operation_id, :context_id);
+            ",
+            named_params! {
+                ":name": &action.name,
+                ":key": &key,
+                ":irmin_time": &action.irmin_time,
+                ":tezedge_time": &action.tezedge_time,
+                ":block_id": block_id,
+                ":operation_id": operation_id,
+                ":context_id": context_id
+            },
+        )?;
+
+        self.nactions = self
+            .nactions
+            .checked_add(1)
+            .expect("actions count overflowed");
+
+        self.times_in_current_block_by_name
+            .entry(action.name.clone())
+            .or_default()
+            .add(action.irmin_time, action.tezedge_time);
+
+        Ok(())
+    }
+
+    // Compute stats for the current block and global ones
+    fn update_global_stats(&mut self) -> Result<(), SQLError> {
```

Comment:
Instead of having to keep a vector of all timings in `*_times`, we can use an SQL query to calculate these values (will probably be slightly slower, but avoids having to keep those vectors) using the aggregate functions:

https://www.sqlite.com/lang_aggfunc.html#sumunc (`total`, `max` and `count`).

Btw, not a big deal, just something that can be improved a bit, but not a priority.